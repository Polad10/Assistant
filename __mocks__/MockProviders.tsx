import { Dispatch, PropsWithChildren, SetStateAction } from 'react'
import { Translator } from '../helpers/Translator'
import { AppointmentRequest } from '../modals/Appointment'
import { PatientRequest } from '../modals/Patient'
import { Setting } from '../modals/Setting'
import { TreatmentRequest } from '../modals/Treatment'
import { PaymentRequest } from '../modals/Payment'
import { Api } from '../helpers/Api'
import { mockModels } from './MockModels'
import { AuthContext } from '../contexts/AuthContext'
import { DataContext } from '../contexts/DataContext'
import { ThemeContext } from '../contexts/ThemeContext'
import { LocalizationContext } from '../contexts/LocalizationContext'

const themeContextMock = {
  primary: '#121212',
  secondary: '#1a1a1a',
  accent: '#7600eb',
  neutral: '#faf8ff',
  border: '#374151',
  info: '#d3fbd8',
  success: '#00e5b0',
  warning: '#9f1239',
  error: '#9f1239',
}

const localizationContextMock = {
  translator: new Translator('en'),
  language: 'en',
}

const authContextMock = {
  user: undefined,
  loading: false,
}

const dataContextMock = {
  fetchPatients: async () => {},
  fetchAppointments: async () => {},
  fetchTreatments: async () => {},
  fetchPayments: async () => {},
  fetchSetting: async () => {},

  createAppointment: async (appointmentRequest: AppointmentRequest) => mockModels.appointment_1,
  updateAppointment: async (appointmentRequest: AppointmentRequest) => mockModels.appointment_1,
  deleteAppointment: async (appointmentId: number) => {},

  createPatient: async (patientRequest: PatientRequest) => mockModels.patient_1,
  updatePatient: async (patientRequest: PatientRequest) => mockModels.patient_1,
  deletePatient: async (patientId: number) => {},

  createTreatment: async (treatmentRequest: TreatmentRequest) => mockModels.treatment_1,
  updateTreatment: async (treatmentRequest: TreatmentRequest) => mockModels.treatment_1,
  deleteTreatment: async (treatmentId: number) => {},

  createPayment: async (paymentRequest: PaymentRequest) => mockModels.payment_1,
  updatePayment: async (paymentRequest: PaymentRequest) => mockModels.payment_1,
  deletePayment: async (paymentId: number) => {},

  createSetting: async (setting: Setting) => {},
  updateSetting: async (setting: Setting) => {},

  api: undefined,
  setApi: (action: SetStateAction<Api | undefined>): Dispatch<SetStateAction<Api | undefined>> => jest.fn(),

  loading: false,
  setLoading: (action: SetStateAction<boolean>): Dispatch<SetStateAction<boolean>> => jest.fn(),

  patients: [mockModels.patient_1, mockModels.patient_2],
  appointments: [mockModels.appointment_1, mockModels.appointment_2],
  treatments: [mockModels.treatment_1, mockModels.treatment_2],
  payments: [mockModels.payment_1, mockModels.payment_2],
  setting: mockModels.setting,
}

export default function MockProviders({ children }: PropsWithChildren) {
  return (
    <AuthContext.Provider value={authContextMock}>
      <DataContext.Provider value={dataContextMock}>
        <ThemeContext.Provider value={themeContextMock}>
          <LocalizationContext.Provider value={localizationContextMock}>{children}</LocalizationContext.Provider>
        </ThemeContext.Provider>
      </DataContext.Provider>
    </AuthContext.Provider>
  )
}
