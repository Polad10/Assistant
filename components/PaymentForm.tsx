import { DeviceEventEmitter, NativeSyntheticEvent, TextInputChangeEventData } from 'react-native'
import MyInput from './MyInput'
import { useNavigation } from '@react-navigation/native'
import { useCallback, useContext, useEffect, useRef, useState } from 'react'
import { RootStackParamList, RootStackScreenProps } from '../types/Navigation'
import HeaderButton from './HeaderButton'
import { Payment, PaymentRequest } from '../modals/Payment'
import { DateTime } from 'luxon'
import { DataContext } from '../contexts/DataContext'
import DateInput, { DateInputRefType } from './DateInput'
import CreateButton from './CreateButton'
import MainView from './MainView'
import { Treatment } from '../modals/Treatment'
import TouchableInput from './TouchableInput'
import TouchableWithoutFeedbackInput from './TouchableWithoutFeedbackInput'
import { Patient } from '../modals/Patient'
import MyKeyboardAvoidingView from './MyKeyboardAvoidingView'
import { FontAwesome6 } from '@expo/vector-icons'
import { ThemeContext } from '../contexts/ThemeContext'
import { translate } from '../helpers/Translator'

type Props = {
  pageName: keyof RootStackParamList
  treatmentId?: number
  payment?: Payment
  patient?: Patient
}

export default function PaymentForm(props: Props) {
  const navigation = useNavigation<RootStackScreenProps<'NewPayment'>['navigation']>()

  const themeContext = useContext(ThemeContext)!
  const context = useContext(DataContext)!

  const treatmentId = props.payment?.treatment_id || props.treatmentId
  const treatment = context.treatments?.find((t) => t.id === treatmentId)

  const treatmentEditable = !treatment

  let dateInitialVal = props.payment ? new Date(props.payment.date) : undefined

  const [amount, setAmount] = useState(props.payment?.amount.toString())
  const [selectedTreatment, setSelectedTreatment] = useState(treatment)

  const [showAmountInputError, setShowAmountInputError] = useState(false)
  const [showDatePickerError, setShowDatePickerError] = useState(false)
  const [showTreatmentInputError, setShowTreatmentInputError] = useState(false)

  const [focusedInputIndex, setFocusedInputIndex] = useState(0)

  const dateInputRef = useRef<DateInputRefType>()

  const handleSave = useCallback(async () => {
    if (validate()) {
      const date = dateInputRef.current?.getDate()

      const newPaymentRequest: PaymentRequest = {
        date: DateTime.fromJSDate(date!).toISODate()!,
        amount: Number(amount),
        treatment_id: selectedTreatment!.id,
      }

      if (props.payment) {
        newPaymentRequest.id = props.payment.id
      }

      DeviceEventEmitter.emit('paymentSaved', newPaymentRequest)
    }
  }, [amount, selectedTreatment])

  function validate() {
    let valid = true

    const date = dateInputRef.current?.getDate()

    if (!date) {
      valid = false
      setShowDatePickerError(true)
    }

    if (!amount || isNaN(Number(amount))) {
      valid = false
      setShowAmountInputError(true)
    }

    if (!selectedTreatment) {
      valid = false
      setShowTreatmentInputError(true)
    }

    return valid
  }

  useEffect(() => {
    if (props.payment) {
      navigation.setOptions({
        headerRight: () => <HeaderButton title={translate('save')} onPress={handleSave} />,
      })
    }
  }, [navigation, handleSave])

  useEffect(() => {
    const treatmentSelectedListener = DeviceEventEmitter.addListener('treatmentSelected', handleTreatmentSelect)
    const treatmentCreatedListener = DeviceEventEmitter.addListener('treatmentCreated', handleTreatmentSelect)

    return () => {
      treatmentSelectedListener.remove()
      treatmentCreatedListener.remove()
    }
  }, [])

  function handleAmountChange(event: NativeSyntheticEvent<TextInputChangeEventData>) {
    setShowAmountInputError(false)
    setAmount(event.nativeEvent.text.replace(',', '.'))
  }

  function handleTreatmentSelect(treatment: Treatment) {
    navigation.navigate('NewPayment', { treatmentId: props.treatmentId, patient: props.patient })

    setShowTreatmentInputError(false)
    setSelectedTreatment(treatment)
  }

  function handleTreatmentChange() {
    if (treatmentEditable) {
      navigation.navigate('Treatments', { patient: props.patient })
    }
  }

  return (
    <MyKeyboardAvoidingView focusedInputIndex={focusedInputIndex}>
      <MainView style={{ paddingTop: 20 }}>
        <DateInput
          ref={dateInputRef}
          label={translate('date')}
          placeholder={translate('pickDate')}
          date={dateInitialVal}
          showError={showDatePickerError}
          onChange={() => setShowDatePickerError(false)}
          onFocus={() => setFocusedInputIndex(0)}
        />
        <MyInput
          label={translate('amount')}
          placeholder={translate('enterAmount')}
          value={amount}
          showError={showAmountInputError}
          onChange={handleAmountChange}
          keyboardType='decimal-pad'
          rightIcon={<FontAwesome6 name='manat-sign' color={themeContext.neutral} size={20} />}
          onFocus={() => setFocusedInputIndex(1)}
        />

        {treatmentEditable ? (
          <TouchableInput
            onPress={handleTreatmentChange}
            label={translate('treatment')}
            placeholder={translate('selectTreatment')}
            value={selectedTreatment?.title}
            showError={showTreatmentInputError}
          />
        ) : (
          <TouchableWithoutFeedbackInput
            label={translate('treatment')}
            placeholder={translate('selectTreatment')}
            value={selectedTreatment?.title}
            showError={showTreatmentInputError}
          />
        )}

        {!props.payment && <CreateButton onPress={handleSave} />}
      </MainView>
    </MyKeyboardAvoidingView>
  )
}
