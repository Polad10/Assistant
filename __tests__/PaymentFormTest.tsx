import renderer from 'react-test-renderer'
import MockProviders from '../__mocks__/MockProviders'
import PaymentForm from '../components/PaymentForm'

describe('PaymentForm', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <MockProviders>
          <PaymentForm pageName='NewPayment' />
        </MockProviders>
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })
})
