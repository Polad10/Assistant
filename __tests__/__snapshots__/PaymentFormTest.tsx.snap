// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaymentForm renders correctly 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "backgroundColor": "#121212",
        "flex": 1,
      },
      {
        "paddingHorizontal": 10,
        "paddingTop": 20,
      },
    ]
  }
>
  <View
    onLayout={[Function]}
    style={
      [
        {
          "flex": 1,
        },
        {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <RCTScrollView
      contentContainerStyle={
        {
          "flexGrow": 1,
        }
      }
      onScroll={[Function]}
      scrollEventThrottle={50}
    >
      <View>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            [
              {
                "backgroundColor": "#121212",
                "flex": 1,
              },
              undefined,
            ]
          }
        >
          <View>
            <View
              style={
                {
                  "paddingHorizontal": 0,
                  "width": "100%",
                }
              }
              testID="RNE__Input__view-wrapper"
            >
              <Text
                style={
                  {
                    "color": "#d3fbd8",
                    "fontSize": 16,
                    "fontWeight": "bold",
                    "marginBottom": 5,
                  }
                }
              >
                Date
              </Text>
              <View
                collapsable={false}
                style={
                  {
                    "alignItems": "center",
                    "borderBottomWidth": 1,
                    "borderColor": "#374151",
                    "borderRadius": 10,
                    "borderWidth": 1,
                    "flexDirection": "row",
                    "marginBottom": 20,
                    "transform": [
                      {
                        "translateX": 0,
                      },
                    ],
                  }
                }
              >
                <TextInput
                  autoCorrect={false}
                  editable={true}
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  placeholder="2024-01-01"
                  placeholderTextColor="#86939e"
                  returnKeyType="done"
                  selectionColor="#faf8ff"
                  showError={false}
                  style={
                    {
                      "color": "#faf8ff",
                      "flex": 1,
                      "fontSize": 18,
                      "minHeight": 40,
                      "paddingHorizontal": 10,
                      "paddingTop": 20,
                      "paddingVertical": 20,
                    }
                  }
                  testID="RNE__Input__text-input"
                  underlineColorAndroid="transparent"
                />
                <View
                  style={
                    {
                      "alignItems": "center",
                      "height": 40,
                      "justifyContent": "center",
                      "marginRight": 10,
                      "marginVertical": 4,
                      "paddingRight": 0,
                    }
                  }
                >
                  <Text />
                </View>
              </View>
              <Text
                style={
                  {
                    "color": "#ff190c",
                    "fontSize": 12,
                    "height": 0,
                    "margin": 0,
                    "padding": 0,
                  }
                }
              />
            </View>
            <Modal
              animationType="none"
              hardwareAccelerated={false}
              isVisible={false}
              onHide={[Function]}
              transparent={true}
              visible={false}
            />
          </View>
          <View
            style={
              {
                "paddingHorizontal": 0,
                "width": "100%",
              }
            }
            testID="RNE__Input__view-wrapper"
          >
            <Text
              style={
                {
                  "color": "#d3fbd8",
                  "fontSize": 16,
                  "fontWeight": "bold",
                  "marginBottom": 5,
                }
              }
            >
              Amount
            </Text>
            <View
              collapsable={false}
              style={
                {
                  "alignItems": "center",
                  "borderBottomWidth": 1,
                  "borderColor": "#374151",
                  "borderRadius": 10,
                  "borderWidth": 1,
                  "flexDirection": "row",
                  "marginBottom": 20,
                  "transform": [
                    {
                      "translateX": 0,
                    },
                  ],
                }
              }
            >
              <TextInput
                autoCorrect={false}
                editable={true}
                keyboardType="decimal-pad"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder="100"
                placeholderTextColor="#86939e"
                returnKeyType="done"
                selectionColor="#faf8ff"
                showError={false}
                style={
                  {
                    "color": "#faf8ff",
                    "flex": 1,
                    "fontSize": 18,
                    "minHeight": 40,
                    "paddingHorizontal": 10,
                    "paddingTop": 20,
                    "paddingVertical": 20,
                  }
                }
                testID="RNE__Input__text-input"
                underlineColorAndroid="transparent"
              />
              <View
                style={
                  {
                    "alignItems": "center",
                    "height": 40,
                    "justifyContent": "center",
                    "marginRight": 10,
                    "marginVertical": 4,
                    "paddingRight": 0,
                  }
                }
              >
                <Text />
              </View>
            </View>
            <Text
              style={
                {
                  "color": "#ff190c",
                  "fontSize": 12,
                  "height": 0,
                  "margin": 0,
                  "padding": 0,
                }
              }
            />
          </View>
          <View
            style={
              {
                "paddingHorizontal": 0,
                "width": "100%",
              }
            }
            testID="RNE__Input__view-wrapper"
          >
            <Text
              style={
                {
                  "color": "#d3fbd8",
                  "fontSize": 16,
                  "fontWeight": "bold",
                  "marginBottom": 5,
                }
              }
            >
              Treatment
            </Text>
            <View
              collapsable={false}
              style={
                {
                  "alignItems": "center",
                  "borderBottomWidth": 1,
                  "borderColor": "#374151",
                  "borderRadius": 10,
                  "borderWidth": 1,
                  "flexDirection": "row",
                  "marginBottom": 20,
                  "transform": [
                    {
                      "translateX": 0,
                    },
                  ],
                }
              }
            >
              <TextInput
                autoCorrect={false}
                editable={true}
                onBlur={[Function]}
                onFocus={[Function]}
                onPress={[Function]}
                placeholder="Select a treatment"
                placeholderTextColor="#86939e"
                returnKeyType="done"
                selectionColor="transparent"
                showError={false}
                showSoftInputOnFocus={false}
                style={
                  {
                    "color": "#faf8ff",
                    "flex": 1,
                    "fontSize": 18,
                    "minHeight": 40,
                    "paddingHorizontal": 10,
                    "paddingTop": 20,
                    "paddingVertical": 20,
                  }
                }
                testID="RNE__Input__text-input"
                underlineColorAndroid="transparent"
              />
              <View
                style={
                  {
                    "alignItems": "center",
                    "height": 40,
                    "justifyContent": "center",
                    "marginRight": 10,
                    "marginVertical": 4,
                    "paddingRight": 0,
                  }
                }
              >
                <Text />
              </View>
            </View>
            <Text
              style={
                {
                  "color": "#ff190c",
                  "fontSize": 12,
                  "height": 0,
                  "margin": 0,
                  "padding": 0,
                }
              }
            />
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>
  <RCTSafeAreaView
    style={
      {
        "marginTop": "auto",
      }
    }
  >
    <View
      style={
        [
          {
            "overflow": "hidden",
          },
          {
            "borderRadius": 2,
          },
          undefined,
          false,
        ]
      }
      testID="RNE_BUTTON_WRAPPER"
    >
      <View
        accessibilityRole="button"
        accessibilityState={
          {
            "busy": false,
            "checked": undefined,
            "disabled": false,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "marginHorizontal": 10,
            "opacity": 1,
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#7600eb",
              "borderColor": "#2089dc",
              "borderRadius": 10,
              "borderWidth": 0,
              "flexDirection": "row",
              "justifyContent": "center",
              "padding": 10,
              "paddingHorizontal": 10,
            }
          }
        >
          <Text
            style={
              {
                "color": "white",
                "fontSize": 18,
                "paddingVertical": 1,
                "textAlign": "center",
              }
            }
          >
            Create
          </Text>
        </View>
      </View>
    </View>
  </RCTSafeAreaView>
</View>
`;
